### PR TITLE
[SYCL] Reuse discarded L0 events in scope of command list

### PR DESCRIPTION
**Pull request by againull**
Patch implements reset and reuse of Level Zero events in scope of cmd list. Same level zero events are reused inside different command lists in scope of the queue, we need new event only to switch between command lists.

The scheme in scope of command list looks like this: 
Operation1 = zeCommantListAppendMemoryCopy (signal event1)

zeCommandListAppendBarrier(wait for event1)
zeCommandListAppendEventReset(event1)
Operation2 = zeCommandListAppendMemoryCopy (signal event2)

zeCommandListAppendBarrier(wait for event2)
zeCommandListAppendEventReset(event2)
Operation3 = zeCommandListAppendMemoryCopy (signal event1)

If we switch to a different command list then we signal new event and insert a barrier into new command list waiting on that event.

CmdList1:
Operation1 = zeCommantListAppendMemoryCopy (signal event1)

zeCommandListAppendBarrier(wait for event1)
zeCommandListAppendEventReset(event1)
zeCommandListAppendSignalEvent(NewEvent)

CmdList2:
zeCommandListAppendBarrier(wait for NewEvent)

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 1]
```diff
@@ -609,7 +611,17 @@ struct pi_command_list_info_t {
...
+  // List of kernels in this command list associated with discarded events.
+  std::list<pi_kernel> Kernels;
+
+  // List of dependent events associated with discarded events in this command
+  // list.
+  std::list<_pi_ze_event_list_t> WaitLists;
+
+  size_t NumDiscardedEvents = 0;
+
+  size_t size() const { return EventList.size() + NumDiscardedEvents; }
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:624]: why do we need to handle-discarded events with completely separate data-structures? Can you extend the 

> std::vector<pi_event> EventList{};

for discarded events be just a special case of that existing list?

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 2]
```diff
@@ -830,6 +844,16 @@ struct _pi_context : _pi_object {
...
 
+struct ze_event {
+  // Level Zero event handle.
+  ze_event_handle_t Handle;
+
+  // Level Zero event pool handle.
+  ze_event_pool_handle_t Pool;
+
+  bool HostVisible = false;
+};
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:855]: Why do you need this data structure? It looks like a subset of a _pi_event, so why not just use the later?

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 1 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.hpp:624]: Whole idea of the patch was avoiding creation/deletion of pi_event objects for discarded events.  That's why it is implemented this way.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 2 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.hpp:855]: Idea of the patch was avoiding creation/deletion of pi_event objects for discarded events.  This structure is lightweight (it contains only two pointers and a flag) while pi_event contains more information which is unnecessary in this case.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:624]: Really? I thought the  main idea is to not create excessive L0 events, which you correctly tell in the title:

> Reuse discarded L0 events in scope of command list

I don't think it is important (at least now) to not create PI events, which are just a "thin" wrapper that was never blamed to be a problem.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:855]: It'd be good to reuse PI events, but I don't like the idea to create parallel structures just for the reason of being more "lightweight" where it wasn't proved to be a problem. Even if we had to go this route then this structure should be re-used by the PI event.

Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: I agree with your concern and removed introduced ze_event data structure and now use pi_event instead.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 3]
```diff
@@ -1310,6 +1355,14 @@ struct _pi_ze_event_list_t {
     this->Length = other.Length;
     return *this;
   }
+
+  _pi_ze_event_list_t(const _pi_ze_event_list_t &other) {
+    this->ZeEventList = other.ZeEventList;
+    this->PiEventList = other.PiEventList;
+    this->Length = other.Length;
+  }
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:1363]: this is physically duplicating the lists of events, right? why is that needed? and don't we end up with the same event being then referenced from multiple lists meaning we need to retain/release them in addition?

Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: These changes are redundant after patch update, reverted.

**Comment by againull** [comment]
@smaslov-intel Could you please review.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 4]
```diff
@@ -1073,6 +1073,95 @@ struct _pi_queue : _pi_object {
...
+  // This barrier guarantees that command list execution starts only after
+  // completion of previous command list which signals aforementioned event. It
+  // allows to reset and reuse same event handles inside all command lists in
+  // scope of the queue. It means that we need 2 reusable events of each type
+  // (host-visible and device-scope) per queue at maximum.
+
+  // This data member keeps track of the last used command list and allows to
+  // handle switch of immediate command lists because immediate command lists
+  // are never closed unlike regular command lists.
+  pi_command_list_ptr_t LastCommandList = CommandListMap.end();
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:1121]:
```suggestion
  pi_command_list_ptr_t LastUsedCommandList = CommandListMap.end();
```

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -96,6 +96,16 @@ static const bool DisableEventsCaching = [] {
   return std::stoi(DisableEventsCachingFlag) != 0;
 }();
 
+// This is an experimental option that allows reset and reuse of uncompleted
+// events in the in-order queue with discard_events property.
+static const bool ReuseDiscardedEvents = [] {
+  const char *ReuseDiscardedEventsFlag =
+      std::getenv("SYCL_PI_LEVEL_ZERO_REUSE_DISCARDED_EVENTS");
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:103]: Please document this env var

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 6]
```diff
@@ -649,6 +659,36 @@ ze_result_t ZeCall::doCall(ze_result_t ZeResult, const char *ZeName,
   if (!(condition))                                                            \
     return error;
 
+pi_result _pi_queue::appendWaitAndResetIfLastEventDiscarded(
+    pi_command_list_ptr_t CommandList) {
+  if (LastCommandEvent && LastCommandEvent->IsDiscarded) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:664]: What if the last command's event is not discarded but previous one is?

Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:664]: If the previous event was discarded it was reset and put into the cache after corresponding command submission.
If the last command is not discarded then it is a new event (we don't take it from cache) and don't reset it after command submission and don't put it to the cache.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 7]
```diff
@@ -667,10 +707,20 @@ inline static pi_result createEventAndAssociateQueue(
...
+  if (*Event == nullptr)
+    PI_CALL(EventCreate(Queue->Context, Queue, ForceHostVisible, Event));
 
   (*Event)->Queue = Queue;
   (*Event)->CommandType = CommandType;
+  (*Event)->IsDiscarded = IsInternal;
+  // Discarded event doesn't own ze_event, it is used by multiple pi_event
+  // objects.
+  if (IsInternal)
+    (*Event)->OwnZeEvent = false;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:723]: But we need to delete the L0 event at the end, how is this achieved?

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 8]
```diff
@@ -1073,6 +1073,95 @@ struct _pi_queue : _pi_object {
...
+  // two times in a row and have to do round-robin between two events. Otherwise
+  // it picks an event from the beginning of the cache and returns it. Event
+  // from the last command is always appended to the end of the list.
+  pi_event getEventFromCache(bool HostVisible);
+
+  // Put pi_event to the cache. Provided pi_event object is not used by
+  // any command but its ZeEvent is used by many pi_event objects.
+  // Commands to wait and reset ZeEvent must be submitted to the queue before
+  // calling this method.
+  pi_result addEventToCache(pi_event Event);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:1143]:
```suggestion
  pi_result addEventToQueueCache(pi_event Event);
```
and rename in pi_context to "addEventToContextCache" (for readability)

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 7 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:723]: We achieve this by releasing all events from the cache at queue destruction. Event in the cache owns L0 event.
https://github.com/intel/llvm/pull/7256/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R3880-R3883

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 9]
```diff
@@ -1334,6 +1427,7 @@ pi_result _pi_context::getAvailableCommandList(
         (!ForcedCmdQueue ||
          *ForcedCmdQueue == CommandBatch.OpenCommandList->second.ZeQueue)) {
       CommandList = CommandBatch.OpenCommandList;
+      PI_CALL(Queue->insertStartBarrierIfDiscardEventsMode(CommandList));
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1430]: why there is no need to call for "insertActiveBarriers" on this path but need this barrier?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 10]
```diff
@@ -1572,8 +1671,15 @@ pi_result _pi_queue::executeCommandList(pi_command_list_ptr_t CommandList,
 
   // The list can be empty if command-list only contains signals of proxy
   // events.
-  if (!CommandList->second.EventList.empty())
+  if (!CommandList->second.EventList.empty() &&
+      this->LastCommandEvent != CommandList->second.EventList.back()) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1675]: What is this extra condition doing?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 11]
```diff
@@ -1678,21 +1784,44 @@ pi_result _pi_queue::executeCommandList(pi_command_list_ptr_t CommandList,
...
+          // If we have in-order queue with discarded events then we want to
+          // treat this event as regular event. We insert a barrier in the next
+          // command list to wait for this event.
+          LastCommandEvent = HostVisibleEvent;
+        } else {
+          // For all other queues treat this as a special event and indicate no
+          // cleanup is needed.
+          PI_CALL(piEventReleaseInternal(HostVisibleEvent));
+          HostVisibleEvent->CleanedUp = true;
+        }
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1798]: can we change to treat it "regular" in all cases?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 12]
```diff
@@ -1938,8 +2080,48 @@ pi_result _pi_ze_event_list_t::createAndRetainPiZeEventList(
...
 
+  if (CurQueue->isInOrderQueue() && CurQueue->LastCommandEvent != nullptr) {
+    if (CurQueue->Device->useImmediateCommandLists()) {
+      if (ReuseDiscardedEvents && CurQueue->isDiscardEvents()) {
+        // If queue is in-order with discarded events and if
+        // new command list is different from the last used command list then
+        // signal new event from the last immediate command list. We are going
+        // to insert a barrier in the new command list waiting for that event.
+        auto QueueGroup = CurQueue->getQueueGroup(UseCopyEngine);
+        auto NextImmCmdList = QueueGroup.ImmCmdLists[QueueGroup.NextIndex];
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2091]: it's fragile to use "NextIndex" directly here, maybe add a boolean to getQueueIndex that we just want to know the index (but not advance it)?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 13]
```diff
@@ -1938,8 +2080,48 @@ pi_result _pi_ze_event_list_t::createAndRetainPiZeEventList(
...
+        if (CurQueue->LastCommandList != CurQueue->CommandListMap.end() &&
+            CurQueue->LastCommandList != NextImmCmdList) {
+          CurQueue->signalEventFromCmdListIfLastEventDiscarded(
+              CurQueue->LastCommandList);
+        }
+      }
+    } else {
+      // Ensure LastCommandEvent's batch is submitted if it is differrent
+      // from the one this command is going to.
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2100]: Please also comment that the "signalEventFromCmdListIfLastEventDiscarded" will be called at batch close if needed.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 9 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1430]: We need to call insertStartBarrierIfDiscardEventsMode here because we need to insert this barrier whenever we start working with new command list (deferent from the last used) -  it doesn't matter if it was recycled or new command list was created or we just took existing open command list. In this branch we take existing open command list, it may be different from the last used command list, for example if there are from different engines.

According to my understanding insertActiveBarriers cares only about new command lists which may be potentiallu created in new L0 queues  where we haven't put  active barriers yet. If recycled or existing open command list is used then active barriers were already inserted on corresponding ze_queue before.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 7 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:723]: OK, please tell it in the source comment here.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 6 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:664]: sounds ok

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 9 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1430]: > According to my understanding insertActiveBarriers cares only about new command lists which may be potentiallu created in new L0 queues where we haven't put active barriers yet. If recycled or existing open command list is used then active barriers were already inserted on corresponding ze_queue before.

Hmm, that's vague. We also have some known problems with "ActiveBarries" so maybe they are related.
I was going to suggest that we do "insertStartBarrierIfDiscardEventsMode" from "insertActiveBarriers".
But OK, let's keep it this way for now.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:103]: Documented.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 14]
```diff
@@ -649,6 +659,36 @@ ze_result_t ZeCall::doCall(ze_result_t ZeResult, const char *ZeName,
...
+    } catch (const std::bad_alloc &) {
+      return PI_ERROR_OUT_OF_HOST_MEMORY;
+    } catch (...) {
+      return PI_ERROR_UNKNOWN;
+    }
+
+    if (LastCommandEvent->isHostVisible())
+      PiEvent->HostVisibleEvent = PiEvent;
+
+    PI_CALL(addEventToCache(PiEvent));
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I agree, renamed.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 7 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:729]: Added a comment.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 10 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1684]: I added a clarifying comment.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 15]
```diff
@@ -1572,8 +1671,15 @@ pi_result _pi_queue::executeCommandList(pi_command_list_ptr_t CommandList,
 
   // The list can be empty if command-list only contains signals of proxy
   // events.
-  if (!CommandList->second.EventList.empty())
+  if (!CommandList->second.EventList.empty() &&
+      this->LastCommandEvent != CommandList->second.EventList.back()) {
     this->LastCommandEvent = CommandList->second.EventList.back();
+    if (ReuseDiscardedEvents && isInOrderQueue() && isDiscardEvents()) {
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Created a helper method.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 11 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Yes, we can, I added TODO for this because it may require test changes as well.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 12 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I agree, fixed as you suggested.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 13 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Added a comment.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 8 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: Good suggestion, renamed.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 16]
```diff
@@ -1938,8 +2080,48 @@ pi_result _pi_ze_event_list_t::createAndRetainPiZeEventList(
...
+
+  bool IncludeLastCommandEvent =
+      CurQueue->isInOrderQueue() && CurQueue->LastCommandEvent != nullptr;
+
+  // If the last event is discarded then we already have a barrier waiting for
+  // that event, so don't need to include the last command event into the wait
+  // list.
+  if (ReuseDiscardedEvents && CurQueue->isDiscardEvents() &&
+      CurQueue->LastCommandEvent->IsDiscarded)
+    IncludeLastCommandEvent = false;
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Yes, this is important, because we reset discarded event, so we can't add it into the wait list -> it will cause waiting for reset event. I added a clarifying comment. Unfortunately, currently I don't see how to get rid of this part.

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Review by pvchupin — APPROVED** [review]

**Comment by againull** [comment]
SYCL :: GroupAlgorithm/SYCL2020/sort.cpp failure on HIP/CUDA is known.
Expected to be fixed by https://github.com/intel/llvm-test-suite/pull/1396
